### PR TITLE
Add topk logits torch op for DS3.2.

### PR DESCRIPTION
**Pull request by dcampora**
## Purpose

This PR adds a CUDA optimized version of logits.topk, and replaces uses of it for both prefill and decode.

## Test

gsm8k 20-shot:

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9583|±  |0.0055|
|     |       |strict-match    |    20|exact_match|↑  |0.9575|±  |0.0056|
```

gpqa-diamond:

```
|       Tasks       |Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-------------------|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gpqa_diamond_n_shot|      2|flexible-extract|     4|exact_match|↑  |0.6768|±  |0.0333|
|                   |       |strict-match    |     4|exact_match|↑  |0.5404|±  |0.0355|
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: csrc/sampler.cu [hunk 1]
```diff
@@ -44,6 +44,236 @@ __global__ void apply_repetition_penalties_kernel(
...
+                                    int* outIndices,
+                                    float* outLogits,
+                                    int stride0,
+                                    int stride1) {
+
+  // The number of bins in the histogram.
+  static constexpr int kNumBins = 512;
+
+  // The top-k width.
+  static constexpr int kTopK = 2048;
```
[the remark this answers is not in the document]
Comment (dcampora) [reply csrc/sampler.cu:65]: topK will be used soon when spec dec support arrives.

[the remark this answers is not in the document]
Comment (dcampora) [reply csrc/sampler.cu:65]: It is safe. We'll support other topK values in a follow-up once spec dec is there.

File: vllm/model_executor/models/deepseek_v2.py [hunk 2]
```diff
@@ -655,26 +660,30 @@ def sparse_attn_indexer(
...
-        # that is out of range(masked already)
-        # this will happen if context length is shorter than K
-        topk_indices[topk_indices > index_end_pos] = -1
+                         next_n_offset + 1).unsqueeze(1)
+        k = min(topk_tokens, logits.shape[-1])
+        num_rows = logits.shape[0]
+        topk_indices = torch.zeros(num_rows,
+                                   topk_tokens,
+                                   dtype=torch.int32,
+                                   device=logits.device)
```
Comment (youkaichao) [inline vllm/model_executor/models/deepseek_v2.py]: do you need to fill it as `-1` during init?

**Review by youkaichao — COMMENTED** [review]
can you add some tests for this kernel?

File: vllm/model_executor/models/deepseek_v2.py (hunk 2 above)
Comment (dcampora) [reply vllm/model_executor/models/deepseek_v2.py]: No need to fill it, the kernel fills the unused elements with `-1`. It could be `torch.empty`, I'll change that.

Comment (dcampora) [reply vllm/model_executor/models/deepseek_v2.py]: I changed it to `torch.empty` now.

**Comment by heheda12345** [comment]
@dcampora can you rebase the PR with main? There is an important CI fix that is needed to run the CI of this PR.

File: tests/kernels/test_top_k_per_row.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    # Run CUDA implementation
+    torch.ops._C.top_k_per_row(logits, row_starts, row_ends, indices, values,
+                               num_rows, top_k, logits.stride(0),
+                               logits.stride(1))
+
+    # Run reference implementation
+    torch_values, torch_indices = logits.topk(min(top_k, max(row_ends)),
+                                              dim=-1)
+    mask_lo = torch_indices >= 0
+    mask_hi = torch_indices - (row_ends - row_starts)[:, None] < 0
```
Comment (youkaichao) [inline tests/kernels/test_top_k_per_row.py]:
```suggestion
    mask_hi = (torch_indices - (row_ends - row_starts)[:, None]) < 0
```

?

File: tests/kernels/test_top_k_per_row.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    mask = torch.full_like(torch_indices,
+                           False,
+                           dtype=torch.bool,
+                           device=torch_indices.device)
```
Comment (youkaichao) [inline tests/kernels/test_top_k_per_row.py]: unnecessary? `mask = mask_lo & mask_hi` creates a new mask

File: tests/kernels/test_top_k_per_row.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        cuda_set = set(cuda_row_indices.tolist())
+        torch_set = set(torch_row_indices.tolist())
+        if cuda_set == torch_set:
+            continue
+
+        # Any difference in elements, compare the values
+        cuda_row_values = cuda_values[row_idx][:num_valid].cpu()
+        torch_row_values = torch_values[row_idx][:num_valid].cpu()
+
+        cuda_values, torch_values = [], []
```
Comment (youkaichao) [inline tests/kernels/test_top_k_per_row.py]: variable shadowing, please rename variables

File: tests/kernels/test_top_k_per_row.py (hunk 3 above)
Comment (youkaichao) [reply tests/kernels/test_top_k_per_row.py]: okay, the order is correct. maybe just add one more parenthesis to be clear.

File: tests/kernels/test_top_k_per_row.py (hunk 4 above)
Comment (dcampora) [reply tests/kernels/test_top_k_per_row.py]: It looks like it. Please note I copied this code from what is now in main. I'll fix it here.

Comment (dcampora) [reply tests/kernels/test_top_k_per_row.py]: Done.

File: tests/kernels/test_top_k_per_row.py (hunk 5 above)
Comment (dcampora) [reply tests/kernels/test_top_k_per_row.py]: Renamed to `cuda_only_values` and `torch_only_values`.

File: csrc/sampler.cu [hunk 6]
```diff
@@ -44,6 +44,245 @@ __global__ void apply_repetition_penalties_kernel(
...
+  // Sort the elements.
+  TopKSort(smemFinal.topKSort)
+      .SortDescendingBlockedToStriped(topKLogits, topKIndices);
+
+// Store to global memory.
+#pragma unroll
+  for (int ii = 0; ii < kNumTopKItemsPerThread; ++ii) {
+    int offset = rowIdx * kTopK + ii * kNumThreadsPerBlock + threadIdx.x;
+    outIndices[offset] = topKIndices[ii];
+    outLogits[offset] = topKLogits[ii];
```
[the remark this answers is not in the document]
Comment (dcampora) [reply csrc/sampler.cu:282]: Removed k parameter, as it is not used.

File: vllm/model_executor/models/deepseek_v2.py [hunk 7]
```diff
@@ -643,17 +643,25 @@ def sparse_attn_indexer(
...
+                chunk.cu_seqlen_ks,
+                chunk.cu_seqlen_ke,
+                topk_indices,
+                topk_values,
+                num_rows,
+                logits.stride(0),
+                logits.stride(1),
+            )
+
+            topk_indices -= chunk.cu_seqlen_ks[:, None]
```
Comment (heheda12345) [inline vllm/model_executor/models/deepseek_v2.py]: For length < 2048 and thus the last few indices are filled with -1, will this cause any problem?

File: csrc/sampler.cu [hunk 8]
```diff
@@ -44,6 +44,245 @@ __global__ void apply_repetition_penalties_kernel(
...
+
+template <int kNumThreadsPerBlock = 512>
+static __global__ void topKPerRow(const float* logits, const int* rowStarts,
+                                  const int* rowEnds, int* outIndices,
+                                  float* outLogits, int stride0, int stride1) {
+  // The number of bins in the histogram.
+  static constexpr int kNumBins = 512;
+
+  // The top-k width.
+  static constexpr int kTopK = 2048;
```
Comment (heheda12345) [inline csrc/sampler.cu:65]: can you add an assert for this kernel only supports 2048 before calling this kernel?

File: vllm/model_executor/models/deepseek_v2.py (hunk 7 above)
Comment (dcampora) [reply vllm/model_executor/models/deepseek_v2.py]: This is the expected behaviour.

File: csrc/sampler.cu (hunk 8 above)
Comment (dcampora) [reply csrc/sampler.cu:65]: Added an assert in the python code before invoking this function, checking that `topk_tokens == 2048`.

File: vllm/model_executor/models/deepseek_v2.py (hunk 7 above)
Comment (heheda12345) [reply vllm/model_executor/models/deepseek_v2.py]: Why is it the expected behavior? `topk_indices` was -1 before this line, and becomes `-1-chunk.cu_seqlen_ks` after this line

Comment (dcampora) [reply vllm/model_executor/models/deepseek_v2.py]: Thanks for spotting this, it should be fixed now.

**Review by heheda12345 — APPROVED** [review]
LGTM! Thank you very much.

**Comment by youkaichao** [comment]
@crazy-JiangDongHua you can continue the discussion in https://github.com/vllm-project/vllm/issues/26554
